### PR TITLE
chores: tiny refactoring

### LINKED ISSUE #2: Cannot override locale :it
Hello there. Thanks for making this gem available; it is very much needed.

I have a question about locales. I see that you set the locale/language to :it in two places in your code; one is in qr-params.rb with `language: "it",` and the other one is in qr-bills.rb, where it is passed as the default locale `I18n.default_locale = :it`. The problem is that I cannot seem to be able to override this locale, unless I comment out the code from qr-bills.rb. Am I missing something?

Ideally, wouldn't the default locale be set in a configuration file, though, instead of the code?

### PR DESCRIPTION
**Pull request by alexanderadam**
This is a very tiny refactoring.
Also this removes setting `I18n.default_locale` (therefore this should fix #2). No gem should ever do that!
Please set locale defaults in your application instead and don't mess with user settings.

Also the Gem modules should be scoped within a proper namespace and you don't have to use instance variables in specs. Use `let` instead. And you don't need `RSpec.configure` in every spec file. Hooks like `before` are available within a `describe` block.
You can see [here an example](https://github.com/damoiser/qr-bills/pull/4/files#diff-9a4b3ecedfeccdf3c19fbcd87f2c2ccb26bab22c116ccc22f76f303466eb2855).
So I guess there are still enough things to do. 😉  But the gem is working, which is great. 👍 

Also you can ask me anytime in case you need feedback or some help.

Apart from the few refactoring things `png` is now supported as a direct output format. 😉

### CODE REVIEW

File: lib/qr-bills.rb [hunk 1]
```diff
@@ -4,41 +4,32 @@
...
-        raise QRExceptions::NOT_SUPPORTED + ": html is the only output format supported so far"
-      end
+      I18n.load_path << locale_file
+    end
 
+    output = case qr_params[:output_params][:format]
+    when 'html'
+      QRHTMLLayout.create(qr_params)
+    when 'png'
+      QRGenerator.create(qr_params, qr_params[:qrcode_filepath])
```
Comment (damoiser) [inline lib/qr-bills.rb:23]: `QRGenerator.create` returns only the qr-code and `QRHTMLLayout.create` return the full qr-bill layout. Seen that the format given here is `html` or `png` I would expect that `png` returns an image of the full qr-bill layout. 

To make this more consistent, I would suggest to use a dedicated `[:output_params][:format]` for the qr-code only, e.g.:

```
 output = case qr_params[:output_params][:format]
    when 'html'
      QRHTMLLayout.create(qr_params)
    when 'qrcode-png'
      QRGenerator.create(qr_params, qr_params[:qrcode_filepath])

    # if in future we'll offer as well full layout based on image
   when 'png'
     QRPNGLayout.create(qr_params)
```

**Review by damoiser — COMMENTED** [review]
thank you @alexanderadam for this improvements as well! I really appreciated your support on this!

I just added a comment, please check it. 

There are some open conflicts given the other merged PRs, can you please take a look?

Thanks again!

File: lib/qr-bills.rb (hunk 1 above)
Comment (alexanderadam) [reply lib/qr-bills.rb:23]: > I would expect that png returns an image of the full qr-bill layout.

Would you consider this as a useful output?
The reason why I stumbled about this gem is because I simply needed a valid QR-Bill-code image.

Is the HTML standardized? If not I would even rather make the HTML part optional and overridable (Ruby has `ERB` for that instead the full blown string that is currently integrated). Thus I would even rather go the opposite way and rather loosen the static HTML thingy. :wink: 

> I would suggest to use a dedicated `[:output_params][:format]` for the qr-code only

Well, I would even remove the whole `get_qr_params` logic anyway. It's not necessary. These are only simple default parameters.
Thus it would be much simpler to add the parameters in a `DEFAULT_PARAMETERS` constant and deep merge all given attributes.
There's really no need to again give empty hashes and values as an argument just to mention that they are empty. :wink:

Comment (damoiser) [reply lib/qr-bills.rb:23]: > Would you consider this as a useful output?
yes :-) it always depends on the use case, you only need the QR Code, someone else like me and some others devs requires to have a html/image qr-bill, how this has been done can be for sure be discussed and as always improved.

This gem is already been used in production (at least by me and probably by someone else), this means that we cannot build changes that breaks the current API / exposed methods.

Back to my original feedback, this is only wording, to stay coherent with the current output format: html, png, jpeg... will produce the full qr-bill (layouts), other kind of output like qrcode-png will return only a subpart of the qr-bill.

Comment (alexanderadam) [reply lib/qr-bills.rb:23]: > > Would you consider this as a useful output?

> yes :-)

Just to clarify: you are saying that it would be useful to have _a full document with text information rendered within a PNG image_?

> This gem is already been used in production (at least by me and probably by someone else), this means that we cannot build changes that breaks the current API / exposed methods.

**TL;DR:** Yes you can. :wink: 

This is actually what version numbers are for. You won't have breaking changes for the same version number but if you are updating a gem (or any other library) to a newer version you definitely might get some breaking changes (there are different version schemes with different rules in which version a breaking change might appear, so you might want to make a version bump in that case).
For instance [the Chunky-PNG example I mentioned before](https://github.com/damoiser/qr-bills/pull/4#issue-606140550) was exactly about this. Chunky PNG simply changed it's API and the compatibility of oily png got broken (hence it needed the PR 21). But if you are using an older version of Chunky PNG you can still use oily png (or if you are using the patched version).
API changes happen very often and are _a very normal_ thing. So nobody's app will be broken if QR bills gem changes its API, since the `Gemfile.lock` in a project directory ensures that the same version will be installed.
So API changes might only happen when you explicitly update a version (i.e. with `bundle update`). But in this case API changes can happen anyway for other gems too. You also can (and should) restrict versions in your `Gemfile`.

Comment (damoiser) [reply lib/qr-bills.rb:23]: > Just to clarify: you are saying that it would be useful to have a full document with text information rendered within a PNG image?

not a full document, only a picture of the QR Bill (QR code and details like address, receipt,.. etc), see attached picture. If you look on the SIX technical requirements (in the readme) there are lot of constraint to be followed (fonts style, size, distances,... etc). This is a pain if you have to do every time. Usually you can simply solve this using the QR HTML Template (the standard output) but you could as well use the png picture who represent the full QR Bill (not active at the moment, but can be a future enhancement), in the same way as the following picture.

image

Yep, I know about versioning :-) here I am trying to follow semantic versioning (https://semver.org/), this means that breaking changes must occurs only on majors versions bump, with your changes I would like to publish the v1 of the gem, however maintaining the promise that the interfaces will still be backward compatible. When will promote this to v2, then yes, we can consider introducing breaking changes.

Comment (alexanderadam) [reply lib/qr-bills.rb:23]: > not a full document, only a picture of the QR Bill (QR code and details like address, receipt,.. etc),

I'm still not convinced about this. The example to have a PNG without any semantic information (i.e. text that could be copied) really doesn't seem to be a sustainable use case for me. I don't see any advantage in this and I don't understand why anyone would prefer a PNG image containing that information rather than having a proper PDF, HTML or whatever document that can contain text.  Raster image formats cannot be properly semantically processed afterwards. All information is lost. The user cannot copy anything from it. Not even the long reference numbers.

This is the reason why I think that raster graphic formats are only useful for the QR code.

I updated the PR accordingly though so that you will be happy. :wink: 
But I'm really curious whether we will see this use case somewhen.

> this means that breaking changes must occurs only on majors versions bump

Sure but currently this gem doesn't have a major version yet, though. Therefore you don't have to promise a stable API.

And personally this gem isn't `1.0.0`-ready yet anyway:

- [ ] the classes should be namespaced within the gem namespace (i.e. `QRCreditorReference` > `QRBills::QRCreditorReference`)
- [ ] [this](https://github.com/alexanderadam/qr-bills/blob/cadc6b9a3af1342bd58722c89f6062b4e00f804a/lib/qr-bills/qr-creditor-reference.rb#L7) should be a constant as well as all the magic numbers (i.e. [`98`](https://github.com/alexanderadam/qr-bills/blob/cadc6b9a3af1342bd58722c89f6062b4e00f804a/lib/qr-bills/qr-creditor-reference.rb#L59) or [`10`](https://github.com/alexanderadam/qr-bills/blob/cadc6b9a3af1342bd58722c89f6062b4e00f804a/lib/qr-bills/qr-creditor-reference.rb#L61))
- [ ] you can remove all these explicit `return`s at the end of a method ([Avoid `return` where not required for flow of control.](https://github.com/rubocop/ruby-style-guide#explicit-return) / [Omit `return` where possible](https://ruby-style-guide.shopify.dev/))
- [ ] [this "html string"](https://github.com/alexanderadam/qr-bills/blob/master/lib/qr-bills/qr-html-layout.rb#L11) should at least be converted to a heredoc with `HTML` delimiter or, even better, a proper ERB template
- [ ] [you wrote some other open TODOs in the source code as well](https://github.com/alexanderadam/qr-bills/blob/cadc6b9a3af1342bd58722c89f6062b4e00f804a/lib/qr-bills/qr-params.rb#L80-L95)
- [ ] [and you wrote even more TODOs in the README](https://github.com/damoiser/qr-bills/blob/9218bf6c244eeb9783da5f317b1d3371a1c9b291/README.md#todo)

> here I am trying to follow semantic versioning (https://semver.org/)

If you would _really_ follow Semantic Versioning then breaking changes _before_ `1.0.0` are actually expected. Only _after_ the first major release the API is considered to be stable.
Also you should release version `1.0.0` if you're thinking that your current API is _actually_ stable (_as mentioned above I wouldn't consider this `1.0.0` read code in this current state, though_).

[Or how semver.org puts this](https://semver.org/#how-do-i-know-when-to-release-100):

> If you have a stable API on which users have come to depend, you should be 1.0.0. If you’re worrying a lot about backwards compatibility, you should probably already be 1.0.0.

So either I missed the major release or breaking changes should be fine according to this. :wink: 
And so far the gem didn't have a major version for most of the time yet. So why would you rush pushing a major version number although you didn't fix your open TODOs and other mentioned points yet?

Comment (damoiser) [reply lib/qr-bills.rb:23]: thanks for the fix and your valid inputs, I mostly agree that are part of the refactoring and improvements. 

However, putting a number into a constant or the improvements mentioned in the readme (except of the "tests improvements" and "improve validation") are not blocking for the v1 version. I never see a business project blocked to going-live because a variable was not typed correctly as a constant or because all features are not been delivered. I am not doing waterfall here where everything needs to be delivered for v1 but thinking more about iteratively improve the component. 

Additionally, I understand that from v0.x to v1.0 we may have breaking changes (but not always expected), from my experiences, breaking changes are accepted only if there is a technical debt to be solved and change of the interface is the only option available, I will only accept them if this is the case. We should try to limit the pain of running behind changes for the developers, this already happens too often and most of the case unnecessary. This gem is not rocket science and we can for sure find valid alternatives :-)

As said, I really appreciate your contributions to this and thanks again for your PRs.

Comment (alexanderadam) [reply lib/qr-bills.rb:23]: > I never see a business project blocked to going-live because a variable was not typed correctly as a constant or because all features are not been delivered.

Me neither but this Gem doesn't even uses a proper namespace. I would really consider the bare minimum.

> I am not doing waterfall here where everything needs to be delivered for v1 but thinking more about iteratively improve the component.

I agree, but labelling a prototype with a stable version number is not useful.
_Please_ explain why you need to put a stable version number on this gem so urgently.
